### ISSUE TITLE
How one can cache Dataset

### ISSUE THREAD

**Issue by REASY** [georust/gdal#425]
Hello, team,

I have a slippy server that serves Slippy Tiles implemented as HTTP server using gdal-rs. Actual rasters are partitioned in many Cloud Optimized GeoTIFF (COG) files with overviews. On high level, I extract tile information from the request that looks like `/:prefix/:layer/:z/:x/:y` and map it to overview and offset to read from COG. My COG files are stored in S3 and I use [vsis3](https://gdal.org/user/virtual_file_systems.html#vsis3). In the beginning of request I open Dataset, in the end it is implicitly closed because of drop. Interestingly, if I query the same slippy tile twice, the only first request has high latency, the second one is much faster (is it because of VSI cache?):
```
2023-07-21T02:13:12.185123Z  INFO tokio-runtime-worker ThreadId(03) qartez_slippy_server::routes: src/routes.rs:169: Read and prepared a tile for .../20/179207/418903.png from /vsis3/.../color/geotiff/5600_13090.tif in 475 ms
2023-07-21T02:13:42.265197Z  INFO tokio-runtime-worker ThreadId(02) qartez_slippy_server::routes: src/routes.rs:169: Read and prepared a tile for .../20/179207/418903.png from /vsis3/.../color/geotiff/5600_13090.tif in 3 ms
```

Does it make sense it such scenario to cache the C descriptor of Dataset and reuse it? Or VSI_CACHE_SIZE together with GDAL_CACHEMAX should be enough? 

Thank you.

**Comment by lnicola** [comment]
Yeah, it's a bit unfortunate. GDAL doesn't allow you to read from a `Dataset` from multiple threads at once, even though cURL could probably support it just fine.

So I think your options are to either:

 - open and close the dataset on each read, which will incur a good bit of overhead (the TLS handshake and and reading the IFDs, I guess)
 - have a thread or pool of threads where each opens the file, gets a read request from a channel, does the actual read, sends the results back, then loops; this should work pretty well, but you'll be storing duplicate data in the GDAL cache

I should probably ask on the mailing list for clarification, though.

**Comment by rouault** [comment]
Starting with GDAL 3.6.0, if the GDAL_NUM_THREADS config option is set, reading in a TIFF/COG file a window of interest that intersects multiple tiles at one will use multithreaded decompression (cf https://github.com/OSGeo/gdal/blob/v3.6.0/NEWS.md), and in GDAL 3.7.0 this was further improved to trigger parallel network requests

**Comment by lnicola** [comment]
I don't think multi-threaded decoding helps in this case (a tile server), since each request will read a single block if everything is set up properly. But we can't have everything just yet :⁠-⁠).

**Comment by ChristianBeilschmidt** [comment]
Isn't the problem that `Dataset`s are not `Send`? You can add Mutexes around it, so that it is `Sync` , but you cannot enforce the `Send`.

There are shared datasets in GDAL, but we haven't implemented them since they cannot simply be used with all the stuff currently implemented for a dataset.

We have done the thread + channel thing that @lnicola mentioned :laughing: .

**Comment by lnicola** [comment]
Yeah, IIRC shared datasets are actually the opposite of the "open the file multiple times" trick. Instead, you (probably) get a mutex around each access, but end up with better cache utilization.

**Comment by rouault** [comment]
> Yeah, IIRC shared datasets are actually the opposite of the "open the file multiple times" trick. Instead, you (probably) get a mutex around each access, but end up with better cache utilization.

no, you don't. You just get the same dataset (if calling GDALOpenShared() from the same thread from which the initial one was opened. Otherwise you'll get a different instance)

**Comment by lnicola** [comment]
Oh, right. Well that's an argument for `Dataset` not being `Send`, because otherwise you can open a shared one twice and pass it to a different thread, which is bad.

**Comment by ChristianBeilschmidt** [comment]
You can't call open `GDALOpenShared` with this library at the moment. This is why we can say that `Dataset: Send`.

There would need to be a second type of `Dataset` , e.g., `SharedDataset`, which would call `GDALOpenShared` under the hood but then not being `Send`.

**Comment by lnicola** [comment]
You're right, there's even a note in the docs:

> Note that the GDAL_OF_SHARED option is removed from the set of allowed option because it subverts the [Send](https://doc.rust-lang.org/nightly/core/marker/trait.Send.html) implementation that allow passing the dataset the another thread. See https://github.com/georust/gdal/issues/154.

**Comment by ManeraKai** [comment]
GDAL isn't thread-safe, but it is re-entrant. So, the following code should be allowed with ranyon:
```rust
  let test: Vec<Geometry> = (0..10).into_par_iter().map(|_| {
        let point =  Geometry::from_wkt("POINT (10, 10)").unwrap();
        point.buffer(3., 1).unwrap()
    }).collect();
```
right?
